File: guests/go/table/table.go
// The table scene (tools/scenes/table.steps): a header click is a REQUEST,
// and the platform sorts nothing.
package table

import (
	"sort"

	kaya "dev.kaya/bindings/go"
)

//go:generate go run dev.kaya/cmd/kaya-gen -type Item -key string
type Item struct {
	Name string
	Size string
}

func App() *kaya.App {
	app := kaya.NewApp()

	// The guest's sort policy; the platform never has one.
	sortedCol := int64(-1)
	sortedDesc := false

	app.Build(func(tx *kaya.Tx) {
		items := ItemCollection(tx)
		// The root is a row: the For's container is the only column.
		var table kaya.Widget
		tx.Mount(tx.Row(func() {
			rows := ItemRows(tx, items).
				Columns([]string{"Name", "Size"}, kaya.SortNone()).
				OnSort(func(tx *kaya.Tx, column uint32) {
					desc := sortedCol == int64(column) && !sortedDesc
					sortedCol, sortedDesc = int64(column), desc
					entries := items.Items(tx)
					sort.SliceStable(entries, func(i, j int) bool {
						a, b := entries[i].Value.Name, entries[j].Value.Name
						if column != 0 {
							a, b = entries[i].Value.Size, entries[j].Value.Size
						}
						if desc {
							return a > b
						}
						return a < b
					})
					// Each key to the end, in the target order.
					for _, e := range entries {
						items.MoveToEnd(tx, e.Key)
					}
					indicator := kaya.SortAsc(column)
					if desc {
						indicator = kaya.SortDesc(column)
					}
					tx.Columns(table, []string{"Name", "Size"}, indicator)
				})
			table = rows.Widget()
			for row := range rows.All() {
				row.Row(func() {
					row.Label(row.Name())
					row.Label(row.Size())
				})
			}
			// Grown on purpose: ungrown, a table hugs its rows.
			tx.SetGrow(table, 1)
		}))
		for _, seed := range []struct{ key, name, size string }{
			{"b", "banana", "30"}, {"a", "apple", "10"}, {"c", "cherry", "20"},
		} {
			items.Insert(tx, seed.key, Item{Name: seed.name, Size: seed.size})
		}
	})

	return app
}

File: guests/java/dev/kaya/milestone2kt/Table.java
package dev.kaya.milestone2kt;

import dev.kaya.KayaApp;
import dev.kaya.KayaGen;
import dev.kaya.KayaRecords;

import java.util.Comparator;
import java.util.List;

/**
 * The table scene from the JVM: column headers and click-to-sort on
 * the For vocabulary (docs/tables-plan.md). A header click is a
 * REQUEST — this guest reorders its collection BY KEY (the reorder
 * scene's idiom) and re-declares the header with the new indicator;
 * the platform sorts nothing. The byte-frozen contract is
 * tools/scenes/table.steps.
 */
final class Table {
    @KayaGen(key = "String")
    record TableItem(String name, String size) {}

    // The guest's sort policy — the platform never has one: clicking
    // the sorted column flips it, clicking another starts ascending.
    private static long sortedCol = -1;
    private static boolean sortedDesc;

    static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            var items = TableItemKaya.collection(tx);
            // The root is a row so the For's container is the scene's
            // only column-kind widget (the reorder scene's rule). The
            // table IS the For, headers declared on the Widget the
            // template form returns.
            tx.mount(tx.row(() -> {
                var table = TableItemKaya.each(tx, items, row -> {
                    row.row(() -> {
                        row.label(row.name);
                        row.label(row.size);
                    });
                });
                // A table is a viewport: like the scroll scene, the
                // scene says it fills — an ungrown table sizes to
                // nothing (measured by screenshot).
                tx.setGrow(table, 1);
                tx.columns(table, new String[] { "Name", "Size" }, KayaApp.Sort.none());
                app.onSort(table, (t, column) -> {
                    boolean desc = sortedCol == column && !sortedDesc;
                    sortedCol = column;
                    sortedDesc = desc;
                    List<KayaRecords.Entry<String, TableItem>> entries = items.items(t);
                    Comparator<KayaRecords.Entry<String, TableItem>> by = column == 0
                        ? Comparator.comparing(e -> e.value.name())
                        : Comparator.comparing(e -> e.value.size());
                    if (desc) {
                        by = by.reversed();
                    }
                    entries.sort(by);
                    // Keys, never indices: moving each key to the end
                    // in the target order leaves the collection sorted.
                    for (var e : entries) {
                        items.moveToEnd(t, e.key);
                    }
                    t.columns(table, new String[] { "Name", "Size" },
                        desc ? KayaApp.Sort.desc(column) : KayaApp.Sort.asc(column));
                });
            }));
            String[][] seeds = { { "b", "banana", "30" }, { "a", "apple", "10" }, { "c", "cherry", "20" } };
            for (String[] seed : seeds) {
                items.insert(tx, seed[0], new TableItem(seed[1], seed[2]));
            }
            return null;
        });

        app.dispatchLoop();
    }

    private Table() {}
}
